### PR TITLE
Refactor of Packet tracking & EOF detection

### PR DESCRIPTION
**Pull request by jonoomph**
Refactor of packet counting and EOF detection. Instead of tracking overall packets read & decoded, we now track per stream (video and audio), so we can more accurately know when a packet is still pending for one of those streams (i.e. async decoding).

Also, refactored variables into a struct, to keep them together (and perhaps, more understandable)

### CODE REVIEW

**Comment by jonoomph** [comment]
[Infinite loop caused by manual override of `has_video` or `has_audio`](https://trello.com/c/xkoOOX4v/172-infinite-loop-caused-by-manual-override-of-hasvideo-or-hasaudio)

File: src/FFmpegReader.cpp [hunk 1]
```diff
@@ -596,15 +590,15 @@ void FFmpegReader::Close() {
...
-		while (packets_decoded < packets_read && attempts < max_attempts) {
+		while (packet_status.packets_decoded() < packet_status.packets_read() && attempts < max_attempts) {
 			ZmqLogger::Instance()->AppendDebugMethod("FFmpegReader::Close (Drain decoder loop)",
-													 "packets_read", packets_read,
-													 "packets_decoded", packets_decoded,
+													 "packets_read", packet_status.packets_read(),
+													 "packets_decoded", packet_status.packets_decoded(),
 													 "attempts", attempts);
-			if (info.has_video) {
+			if (packet_status.video_decoded < packet_status.video_read) {
```
Comment (JacksonRG) [inline src/FFmpegReader.cpp:598]: looks like this will save some compute time :+1:

File: src/FFmpegReader.cpp [hunk 2]
```diff
@@ -991,44 +978,51 @@ std::shared_ptr<Frame> FFmpegReader::ReadStream(int64_t requested_frame) {
...
 			packet = NULL;
-			packets_decoded++;
 		}
 
 		// Determine end-of-stream (waiting until final decoder threads finish)
 		// Force end-of-stream in some situations
-		end_of_file = packets_eof && video_eof && audio_eof;
-		if ((packets_eof && packets_read == packets_decoded) || end_of_file) {
+		packet_status.end_of_file = packet_status.packets_eof && packet_status.video_eof && packet_status.audio_eof;
+		if ((packet_status.packets_eof && packet_status.packets_read() == packet_status.packets_decoded()) || packet_status.end_of_file) {
```
Comment (JacksonRG) [inline src/FFmpegReader.cpp:997]: Looks like this should always break, since we decode as long as there are packets read.

**Review by JacksonRG — APPROVED** [review]
LGTM. I was confused by, but then understood why we took away the hardcoded limit on maximum attempts.

**Pull request merged**